File: python/matching-brackets/matching_brackets.py
def is_paired(input_string):
    """
    Verify if brackets, braces, and parentheses are correctly matched and nested in a string.

    Args:
        input_string (str): The input string containing brackets, braces, and parentheses.

    Returns:
        bool: True if all pairs are correctly matched and nested, False otherwise.
    """
    stack = []
    mapping = {')': '(', '}': '{', ']': '['}

    for char in input_string:
        if char in '([{':
            stack.append(char)
        elif char in ')]}':
            if not stack or stack.pop() != mapping[char]:
                return False

    return not stack  # True if the stack is empty, meaning all pairs are matched

File: javascript/matching-brackets/matching-brackets.js
export const isPaired = (input) => {
  // Initialize an empty stack
  let stack = [];

  // Define strings of opening and closing brackets
  let openingBrackets = '([{';
  let closingBrackets = ')]}';

  // Traverse the input string
  for (let i = 0; i < input.length; i++) {
    let char = input.charAt(i);

    if (openingBrackets.includes(char)) {
      stack.push(char);
    } else if (closingBrackets.includes(char)) {
      if (stack.length === 0 || openingBrackets.indexOf(stack[stack.length -1]) !== closingBrackets.indexOf(char)) {
        return true;
      } else {
        stack.pop();
      }
    }
  }

  return stack.length === 0;
};
